storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +64,395 @@
+            consumer.assign(remoteLogPartitions);
+            this.assignedMetadataPartitions = Collections.unmodifiableSet(metadataPartitionSnapshot);
+            // for newly assigned user-partitions, read from the beginning of the corresponding metadata partition
+            final Set<TopicPartition> seekToBeginOffsetPartitions = assignedUserTopicIdPartitionsSnapshot
+                .stream()
+                .filter(utp -> !utp.isAssigned)
+                .map(utp -> toRemoteLogPartition(utp.metadataPartition))
+                .collect(Collectors.toSet());
+            consumer.seekToBeginning(seekToBeginOffsetPartitions);
+            // for other metadata partitions, read from the offset where the processing left last time.
+            remoteLogPartitions.stream()
+                .filter(tp -> !seekToBeginOffsetPartitions.contains(tp) &&
+                    readOffsetsByMetadataPartition.containsKey(tp.partition()))
+                .forEach(tp -> consumer.seek(tp, readOffsetsByMetadataPartition.get(tp.partition())));
+            // mark all the user-topic-partitions as assigned to the consumer.
+            assignedUserTopicIdPartitionsSnapshot.forEach(utp -> {
+                if (!utp.isAssigned) {
+                    // Note that there can be a race between `remove` and `add` partition assignment. Calling the
+                    // `maybeLoadPartition` here again to be sure that the partition gets loaded on the handler.
+                    remotePartitionMetadataEventHandler.maybeLoadPartition(utp.topicIdPartition);
+                    utp.isAssigned = true;
+                }
+            });
+            processedAssignmentOfUserTopicIdPartitions = assignedUserTopicIdPartitionsSnapshot.stream()
+                .map(utp -> utp.topicIdPartition).collect(Collectors.toSet());
+            clearResourcesForUnassignedUserTopicPartitions(assignedUserTopicIdPartitionsSnapshot);
+            isAllUserTopicPartitionsInitialized = false;
+            uninitializedAt = time.milliseconds();
+            fetchStartAndEndOffsets();
         }
     }
 
-    private void executeReassignment(Set<Integer> assignedMetaPartitionsSnapshot) {
-        Set<TopicPartition> assignedMetaTopicPartitions =
-                assignedMetaPartitionsSnapshot.stream()
-                                              .map(partitionNum -> new TopicPartition(REMOTE_LOG_METADATA_TOPIC_NAME, partitionNum))
-                                              .collect(Collectors.toSet());
-        log.info("Reassigning partitions to consumer task [{}]", assignedMetaTopicPartitions);
-        consumer.assign(assignedMetaTopicPartitions);
+    private void clearResourcesForUnassignedUserTopicPartitions(Set<UserTopicIdPartition> assignedUTPs) {
+        Set<TopicIdPartition> assignedPartitions = assignedUTPs.stream()
+            .map(utp -> utp.topicIdPartition).collect(Collectors.toSet());
```

Comment:
A follow up from comment above
```suggestion
    private void clearResourcesForUnassignedUserTopicPartitions(Set<TopicIdPartition> assignedPartitions) {
```

Reply:
Done